File: redhawk/codegen/jinja/cpp/component/pull/templates/resource_base.cpp
/*#
 * This file is protected by Copyright. Please refer to the COPYRIGHT file 
 * distributed with this source distribution.
 * 
 * This file is part of REDHAWK core.
 * 
 * REDHAWK core is free software: you can redistribute it and/or modify it 
 * under the terms of the GNU Lesser General Public License as published by the 
 * Free Software Foundation, either version 3 of the License, or (at your 
 * option) any later version.
 * 
 * REDHAWK core is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License 
 * for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License 
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 #*/
//% set baseClass = component.superclasses[0].name
//% set artifactType = component.artifacttype
//% set className = component.baseclass.name
/*{% block license %}*/
/*# Allow child templates to include license #*/
/*{% endblock %}*/

/*{% block includes %}*/
#include "${component.baseclass.header}"
/*{% endblock %}*/
/*{% block includeExtentions %}*/
/*# Allow for child template extensions #*/
/*{% endblock %}*/

/*{% block headerComment %}*/
/*******************************************************************************************

    AUTO-GENERATED CODE. DO NOT MODIFY

    The following class functions are for the base class for the ${artifactType} class. To
    customize any of these functions, do not modify them here. Instead, overload them
    on the child class

******************************************************************************************/
/*{% endblock %}*/

/*{% if component is device %}*/
${className}::${className}(char *devMgr_ior, char *id, char *lbl, char *sftwrPrfl) :
    ${baseClass}(devMgr_ior, id, lbl, sftwrPrfl),
/*{% if component is aggregatedevice %}*/
    AggregateDevice_impl(),
/*{% endif %}*/
    serviceThread(0)
{
    construct();
}

${className}::${className}(char *devMgr_ior, char *id, char *lbl, char *sftwrPrfl, char *compDev) :
    ${baseClass}(devMgr_ior, id, lbl, sftwrPrfl, compDev),
/*{% if component is aggregatedevice %}*/
    AggregateDevice_impl(),
/*{% endif %}*/
    serviceThread(0)
{
    construct();
}

${className}::${className}(char *devMgr_ior, char *id, char *lbl, char *sftwrPrfl, CF::Properties capacities) :
    ${baseClass}(devMgr_ior, id, lbl, sftwrPrfl, capacities),
/*{% if component is aggregatedevice %}*/
    AggregateDevice_impl(),
/*{% endif %}*/
    serviceThread(0)
{
    construct();
}

${className}::${className}(char *devMgr_ior, char *id, char *lbl, char *sftwrPrfl, CF::Properties capacities, char *compDev) :
    ${baseClass}(devMgr_ior, id, lbl, sftwrPrfl, capacities, compDev),
/*{% if component is aggregatedevice %}*/
    AggregateDevice_impl(),
/*{% endif %}*/
    serviceThread(0)
{
    construct();
}
/*{% else %}*/
/*{% block componentConstructor %}*/
${className}::${className}(const char *uuid, const char *label) :
    ${baseClass}(uuid, label),
    serviceThread(0)
{
    construct();
}
/*{% endblock %}*/
/*{% endif %}*/

/*{% block construct %}*/
void ${className}::construct()
{
    Resource_impl::_started = false;
    loadProperties();
    serviceThread = 0;
    
    PortableServer::ObjectId_var oid;
/*{% for port in component.ports %}*/
    ${port.cppname} = new ${port.constructor};
    oid = ossie::corba::RootPOA()->activate_object(${port.name});
/*{%   if port.name == 'propEvent' %}*/
/*{%     for property in component.events %}*/
    ${port.cppname}->registerProperty(this->_identifier, this->naming_service_name, this->getPropertyFromId("${property.identifier}"));
/*{%     endfor %}*/
    this->registerPropertyChangePort(${port.cppname});
/*{%   endif %}*/
/*{%   if loop.last %}*/

/*{%   endif %}*/
/*{% endfor %}*/
/*{% for port in component.ports if port is provides %}*/
    registerInPort(${port.cppname});
/*{% endfor %}*/
/*{% for port in component.ports if port is uses %}*/
    registerOutPort(${port.cppname}, ${port.cppname}->_this());
/*{% endfor %}*/
}
/*{% endblock %}*/

/*******************************************************************************************
    Framework-level functions
    These functions are generally called by the framework to perform housekeeping.
*******************************************************************************************/
/*{% block initialize %}*/
void ${className}::initialize() throw (CF::LifeCycle::InitializeError, CORBA::SystemException)
{
}
/*{% endblock %}*/

/*{% block start %}*/
void ${className}::start() throw (CORBA::SystemException, CF::Resource::StartError)
{
    boost::mutex::scoped_lock lock(serviceThreadLock);
    if (serviceThread == 0) {
/*{% for port in component.ports if port.start %}*/
        ${port.cppname}->${port.start};
/*{% endfor %}*/
        serviceThread = new ProcessThread<${className}>(this, 0.1);
        serviceThread->start();
    }
    
    if (!Resource_impl::started()) {
    	Resource_impl::start();
    }
}
/*{% endblock %}*/

/*{% block stop %}*/
void ${className}::stop() throw (CORBA::SystemException, CF::Resource::StopError)
{
    boost::mutex::scoped_lock lock(serviceThreadLock);
    // release the child thread (if it exists)
    if (serviceThread != 0) {
/*{% for port in component.ports if port.stop %}*/
        ${port.cppname}->${port.stop};
/*{% endfor %}*/
        if (!serviceThread->release(2)) {
            throw CF::Resource::StopError(CF::CF_NOTSET, "Processing thread did not die");
        }
        serviceThread = 0;
    }
    
    if (Resource_impl::started()) {
    	Resource_impl::stop();
    }
}
/*{% endblock %}*/

/*{% block getPort %}*/
/*{% if component.ports %}*/
CORBA::Object_ptr ${className}::getPort(const char* _id) throw (CORBA::SystemException, CF::PortSupplier::UnknownPort)
{

    std::map<std::string, Port_Provides_base_impl *>::iterator p_in = inPorts.find(std::string(_id));
    if (p_in != inPorts.end()) {
/*{% for port in component.ports if port is provides %}*/
        if (!strcmp(_id,"${port.name}")) {
            ${port.cpptype} *ptr = dynamic_cast<${port.cpptype} *>(p_in->second);
            if (ptr) {
                return ptr->_this();
            }
        }
/*{% endfor %}*/
    }

    std::map<std::string, CF::Port_var>::iterator p_out = outPorts_var.find(std::string(_id));
    if (p_out != outPorts_var.end()) {
        return CF::Port::_duplicate(p_out->second);
    }

    throw (CF::PortSupplier::UnknownPort());
}

/*{% endif %}*/
/*{% endblock %}*/
/*{% block releaseObject %}*/
void ${className}::releaseObject() throw (CORBA::SystemException, CF::LifeCycle::ReleaseError)
{
    // This function clears the ${artifactType} running condition so main shuts down everything
    try {
        stop();
    } catch (CF::Resource::StopError& ex) {
        // TODO - this should probably be logged instead of ignored
    }

    // deactivate ports
    releaseInPorts();
    releaseOutPorts();

/*{% for port in component.ports %}*/
    delete(${port.cppname});
/*{% endfor %}*/

    ${baseClass}::releaseObject();
}
/*{% endblock %}*/

/*{% block loadProperties %}*/
/*{% from "properties/properties.cpp" import addproperty, initsequence %}*/
void ${className}::loadProperties()
{
/*{% for prop in component.properties %}*/
//% if prop.cppvalues
    ${initsequence(prop)|indent(4)}
//% endif
    ${addproperty(prop)|indent(4)}

/*{% endfor %}*/
}
/*{% endblock %}*/

/*{% block extensions %}*/
/*# Allow for child class extensions #*/
/*{% endblock %}*/

File: redhawk/codegen/jinja/java/component/pull/templates/resource_base.java
/*#
 * This file is protected by Copyright. Please refer to the COPYRIGHT file 
 * distributed with this source distribution.
 * 
 * This file is part of REDHAWK core.
 * 
 * REDHAWK core is free software: you can redistribute it and/or modify it 
 * under the terms of the GNU Lesser General Public License as published by the 
 * Free Software Foundation, either version 3 of the License, or (at your 
 * option) any later version.
 * 
 * REDHAWK core is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License 
 * for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License 
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 #*/
/*{% block license %}*/
/*# Allow child templates to include license #*/
/*{% endblock %}*/
//% set userclass = component.userclass.name
//% set classname = component.baseclass.name
//% set superClass = component.superclass.name
//% set artifactType = component.artifacttype
package ${component.package};

/*{% if component is device %}*/
import java.util.HashMap;
/*{% endif %}*/
import java.util.Properties;

import org.apache.log4j.Logger;

/*{% if component.events %}*/
import org.omg.CORBA.ORB;
/*{% endif %}*/
/*{% if component is not device %}*/
import org.omg.CosNaming.NamingContextPackage.CannotProceed;
import org.omg.CosNaming.NamingContextPackage.InvalidName;
import org.omg.CosNaming.NamingContextPackage.NotFound;
/*{% endif %}*/
/*{% if component.events %}*/
import org.omg.PortableServer.POA;
/*{% endif %}*/
import org.omg.PortableServer.POAPackage.ServantNotActive;
import org.omg.PortableServer.POAPackage.WrongPolicy;

/*{% if component is device %}*/
import CF.DevicePackage.AdminType;
import CF.DevicePackage.OperationalType;
import CF.DevicePackage.UsageType;
/*{% endif %}*/
import CF.InvalidObjectReference;
/*{% filter lines|unique|join('\n') %}*/
/*{%    for prop in component.properties %}*/
/*{%        if prop.CFType %}*/
import CF.${prop.CFType};
/*{%        endif %}*/
/*{%    endfor %}*/
/*{% endfilter %}*/

import org.ossie.component.*;
/*{% if component.properties %}*/
import org.ossie.properties.*;
/*{% endif %}*/
/*{% filter lines|unique|join('\n') %}*/
/*{%   for portgen in component.portgenerators %}*/
/*{%     if loop.first %}*/

/*{%     endif %}*/
/*{%     if portgen.package %}*/
import ${portgen.package}.${portgen.className()};
/*{%     endif %}*/

/*{%   endfor %}*/
/*{% endfilter %}*/

/**
 * This is the ${artifactType} code. This file contains all the access points
 * you need to use to be able to access all input and output ports,
 * respond to incoming data, and perform general ${artifactType} housekeeping
 *
 * Source: ${component.profile.spd}
 *
 * @generated
 */
public abstract class ${classname} extends ${superClass} implements Runnable {
    /**
     * @generated
     */
    public final static Logger logger = Logger.getLogger(${classname}.class.getName());

    /**
     * Return values for service function.
     */
    public final static int FINISH = -1;
    public final static int NOOP   = 0;
    public final static int NORMAL = 1;

/*{% import "base/properties.java" as properties with context %}*/
/*{% for prop in component.properties %}*/
/*{%   if prop is simple %}*/
    ${properties.simple(prop)|indent(4)}
/*{%   elif prop is simplesequence %}*/
    ${properties.simplesequence(prop)|indent(4)}
/*{%   elif prop is struct %}*/
    ${properties.struct(prop)|indent(4)}
/*{%   elif prop is structsequence %}*/
    ${properties.structsequence(prop)|indent(4)}
/*{%   endif %}*/

/*{% endfor %}*/
    // Provides/inputs
/*{% for port in component.ports if port is provides %}*/
    /**
     * @generated
     */
    public ${port.javatype} ${port.javaname};

/*{% endfor %}*/
    // Uses/outputs
/*{% for port in component.ports if port is uses %}*/
    /**
     * @generated
     */
    public ${port.javatype} ${port.javaname};

/*{% endfor %}*/
    /**
     * @generated
     */
    public ${classname}()
    {
        super();
/*{% if component is device %}*/
        this.usageState = UsageType.IDLE;
        this.operationState = OperationalType.ENABLED;
        this.adminState = AdminType.UNLOCKED;
        this.callbacks = new HashMap<String, AllocCapacity>();
/*{% endif %}*/
/*{% for prop in component.properties if prop is structsequence %}*/
        ${properties.structsequence_init(prop)|indent(8)}
/*{% endfor %}*/
/*{% for prop in component.properties %}*/
        addProperty(${prop.javaname});
/*{% endfor %}*/

        // Provides/input
/*{% for port in component.ports if port is provides %}*/
        this.${port.javaname} = new ${port.constructor};
        this.addPort("${port.name}", this.${port.javaname});
/*{% endfor %}*/

        // Uses/output
/*{% for port in component.ports if port is uses %}*/
        this.${port.javaname} = new ${port.constructor};
        this.addPort("${port.name}", this.${port.javaname});
/*{% endfor %}*/
    }

/*{% if component.events %}*/
    /**
     * @generated
     */
    public CF.Resource setup(final String compId, final String compName, final ORB orb, final POA poa) throws ServantNotActive, WrongPolicy
    {
    	CF.Resource retval = super.setup(compId, compName, orb, poa);
    	
/*{% for prop in component.events %}*/
        this.port_propEvent.registerProperty(this.compId, this.compName, this.${prop.javaname});
/*{% endfor %}*/
        
        this.registerPropertyChangePort(this.port_propEvent);
        
    	return retval;
    }

/*{% endif %}*/
    public void start() throws CF.ResourcePackage.StartError
    {
/*{% for port in component.ports if port.start %}*/
        this.${port.javaname}.${port.start};
/*{% endfor %}*/
        super.start();
    }

    public void stop() throws CF.ResourcePackage.StopError
    {
/*{% for port in component.ports if port.stop %}*/
        this.${port.javaname}.${port.stop};
/*{% endfor %}*/
        super.stop();
    }

    public void run() 
    {
        while(this.started())
        {
            int state = this.serviceFunction();
            if (state == NOOP) {
                try {
                    Thread.sleep(100);
                } catch (InterruptedException e) {
                    break;
                }
            } else if (state == FINISH) {
                return;
            }
        }
    }

    protected abstract int serviceFunction();

    /**
     * The main function of your ${artifactType}.  If no args are provided, then the
     * CORBA object is not bound to an SCA Domain or NamingService and can
     * be run as a standard Java application.
     * 
     * @param args
     * @generated
     */
    public static void main(String[] args) 
    {
        final Properties orbProps = new Properties();
        ${userclass}.configureOrb(orbProps);

        try {
            ${superClass}.start_${artifactType}(${userclass}.class, args, orbProps);
        } catch (InvalidObjectReference e) {
            e.printStackTrace();
/*{% if component is not device %}*/
        } catch (NotFound e) {
            e.printStackTrace();
        } catch (CannotProceed e) {
            e.printStackTrace();
        } catch (InvalidName e) {
            e.printStackTrace();
/*{% endif %}*/
        } catch (ServantNotActive e) {
            e.printStackTrace();
        } catch (WrongPolicy e) {
            e.printStackTrace();
        } catch (InstantiationException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }
    }
}
